spec/lib/puppet_spec/ssl.rb

```diff
@@ -0,0 +1,206 @@
+        extension = ext_factory.create_extension(*ext)
+        cert.add_extension(extension)
+      end
+
+      cert.sign(key, DEFAULT_SIGNING_DIGEST)
+
+      cert
+    end
+
+    def self.create_csr(key, name)
+      csr = OpenSSL::X509::Request.new
+
+      csr.public_key  = key.public_key
+      csr.subject     = OpenSSL::X509::Name.parse(name)
+      csr.version     = 2
+      csr.sign(key, DEFAULT_SIGNING_DIGEST)
+
+      csr
+    end
+
+    def self.sign(ca_key, ca_cert, csr, extensions = NODE_EXTENSIONS)
+      cert = OpenSSL::X509::Certificate.new
+
+      cert.public_key = csr.public_key
+      cert.subject    = csr.subject
+      cert.issuer     = ca_cert.subject
+      cert.version    = 2
+      cert.serial     = rand(2**128)
+
+      not_before = just_now
+      cert.not_before = not_before
+      cert.not_after  = not_before + FIVE_YEARS
+
+      ext_factory = extension_factory_for(ca_cert, cert)
+      extensions.each do |ext|
+        extension = ext_factory.create_extension(*ext)
+        cert.add_extension(extension)
+      end
+
+      cert.sign(ca_key, DEFAULT_SIGNING_DIGEST)
+
+      cert
+    end
+
+    def self.create_crl_for(ca_cert, ca_key)
+      crl = OpenSSL::X509::CRL.new
+      crl.version = 1
+      crl.issuer = ca_cert.subject
+
+      ef = extension_factory_for(ca_cert)
+      crl.add_extension(
+        ef.create_extension(["authorityKeyIdentifier", "keyid:always", false]))
+      crl.add_extension(
+        OpenSSL::X509::Extension.new("crlNumber", OpenSSL::ASN1::Integer(0)))
+
+      not_before = just_now
+      crl.last_update = not_before
+      crl.next_update = not_before + FIVE_YEARS
+      crl.sign(ca_key, DEFAULT_SIGNING_DIGEST)
+
+      crl
+    end
+
+    def self.revoke(serial, crl, ca_key)
+      revoked = OpenSSL::X509::Revoked.new
+      revoked.serial = serial
+      revoked.time = Time.now
+      revoked.add_extension(
+        OpenSSL::X509::Extension.new("CRLReason",
+                                     OpenSSL::ASN1::Enumerated(DEFAULT_REVOCATION_REASON)))
+
+      crl.add_revoked(revoked)
+      extensions = crl.extensions.group_by{|e| e.oid == 'crlNumber' }
+      crl_number = extensions[true].first
+      unchanged_exts = extensions[false]
+
+      next_crl_number = crl_number.value.to_i + 1
+      new_crl_number_ext = OpenSSL::X509::Extension.new("crlNumber",
+                                                        OpenSSL::ASN1::Integer(next_crl_number))
+
+      crl.extensions = unchanged_exts + [new_crl_number_ext]
+      crl.sign(ca_key, DEFAULT_SIGNING_DIGEST)
+
+      crl
+    end
+
+    # Creates a self-signed root ca, then signs two node certs, revoking one of them.
+    # Creates an intermediate CA and one node cert off of it.
+    # Creates a leaf CA off of the intermediate CA, then signs two node certs revoking one of them.
+    # Revokes the intermediate CA.
+    # Returns the ca bundle, crl chain, and all the node certs
+    def self.create_chained_pki
```

Comment:
I don'tt know if it's worth it, but this doesn't lend itself very well to using let in rspec which is preferred over instance variables. An alternative might be making each stanza of this into its own function and then you have several lets which refer to the earlier lets. This is definitely not blocking feedback.

Reply:
+1

Reply:
I've been noodling on this a little bit and it I think our use case doesn't lend itself to doing this well with `let`.

1. We need to write either the cert bundle or the cert bundle and crl chain before tests and it feels weird to use let to write to disk (though we can put that into a before block)
2. We need to reference a ca's key, cert, and crl to sign and revoke, so there's only one line that could be removed from the ca stanzas if we used lets for them all
3. We need to have all of our nodes revoked prior to writing out our crl chain. Which means we end up doing a lot of throw away work (creating interim crls) in `let`s and/or using `let!` to force evaluation.

After doing that I was still getting problems... so, honestly I gave up as it no longer seemed worth it. I'd be curious, if this was my lack of understanding of structuring lets, vs our actual use case, but I doubt getting clarity on that is worth the time either. fwiw, here's a commit after I'd gotten a little frustrated and was trying random things: https://github.com/justinstoller/puppet/commit/58077c269600e1514c02b892950866cf88681638